geoips/create_plugin_registries.py

```diff
@@ -14,10 +14,169 @@
+from geoips.errors import PluginRegistryError
+
 
 LOG = logging.getLogger(__name__)
 
 
+def remove_registries(plugin_packages):
+    """Remove all plugin registries if a PluginRegistryError is raised.
+
+    Parameters
+    ----------
+    plugin_packages: list EntryPoints
+        A list of EntryPoints pointing to each installed
+        GeoIPS package --> ie.
+        [EntryPoint(name='geoips', value='geoips', group='geoips.plugin_packages'), ...]
+    """
+    from os import remove
+    from os.path import exists
+
+    LOG.interactive(
+        "\n\n\n\nERROR: Removing registries due to duplicate plugins. You must fix the "
+        "error shown below before GeoIPS can operate correctly.\n"
+        "Once fixed, please run 'create_plugin_registries' to set up GeoIPS "
+        "appropriately\n\n\n"
+    )
+    for pkg in plugin_packages:
+        reg_plug_path = str(resources.files(pkg.value) / "registered_plugins.yaml")
+        if exists(reg_plug_path):
+            remove(reg_plug_path)
+
+
+def registry_sanity_check(plugin_packages):
+    """Check that each plugin package registry has no duplicate lowest depth entries.
+
+    If it does, raise a PluginRegistryError for that specific package, then remove all
+    plugin registries from each package so the user must fix the error before
+    continuing. While this doesn't cause a normal error, duplicate plugins will be
+    overwritten by same-named plugin found in the last package-entrypoint.
+
+    Parameters
+    ----------
+    plugin_packages: list EntryPoints
+        A list of EntryPoints pointing to each installed
+        GeoIPS package --> ie.
+        [EntryPoint(name='geoips', value='geoips', group='geoips.plugin_packages'), ...]
+    """
+    for comp_idx, comp_pkg in enumerate(plugin_packages):
+        # comp_pkg is the package being compared against. This package is compared
+        # against every other available GeoIPS package that is installed.
+        comp_registry = yaml.safe_load(
+            open(resources.files(comp_pkg.value) / "registered_plugins.yaml")
+        )
+        for pkg_idx, pkg in enumerate(plugin_packages):
+            # pkg is the package being compared against comp_pkg. For example, if
+            # comp_pkg was 'geoips', then it would compare against recenter_tc,
+            # data_fusion, template_basic_plugin, etc.
+
+            # The if statement below checks the index of pkg_idx and comp_idx.
+            # If pkg_idk <= comp_idx, that means it's either the same package as
+            # comp_pkg, or that the comparison has already been performed.
+            if pkg_idx <= comp_idx:
+                continue
+            # Track sets of plugins by plugin type
+            # (schemas, yaml_based, and module_based)
+            pkg_registry = yaml.safe_load(
+                open(resources.files(pkg.value) / "registered_plugins.yaml")
+            )
+            for plugin_type in ["yaml_based", "module_based"]:
```

Comment:
Can this list be derived from the registry files rather than hard-coded? We're likely going to add "text_based" in the near future and it is possible that we'll have others farther down the line, too. This would be one less thing to change if we can make it dynamic.

Reply:
I've changed this to dynamically detect this.

Reply:
Looks good. Thanks Evan.